lisa/src/test/java/it/unive/lisa/test/imp/tutorial/ConstantPropagation.java

```diff
@@ -0,0 +1,224 @@
+import it.unive.lisa.symbolic.value.UnaryOperator;
+
+public class ConstantPropagation extends BaseNonRelationalValueDomain<ConstantPropagation> {
+
+	public static final ConstantPropagation TOP = new ConstantPropagation(true, false);
+
+	private static final ConstantPropagation BOTTOM = new ConstantPropagation(false, true);
+
+	private final boolean isTop, isBottom;
+
+	private final Integer value;
+
+	public ConstantPropagation() {
+		this(null, true, false);
+	}
+
+	private ConstantPropagation(Integer value, boolean isTop, boolean isBottom) {
+		this.value = value;
+		this.isTop = isTop;
+		this.isBottom = isBottom;
+	}
+
+	private ConstantPropagation(Integer value) {
+		this(value, false, false);
+	}
+
+	private ConstantPropagation(boolean isTop, boolean isBottom) {
+		this(null, isTop, isBottom);
+	}
+
+	@Override
+	public ConstantPropagation top() {
+		return TOP;
+	}
+
+	@Override
+	public boolean isTop() {
+		return isTop;
+	}
+
+	@Override
+	public ConstantPropagation bottom() {
+		return BOTTOM;
+	}
+
+	@Override
+	public boolean isBottom() {
+		return isBottom;
+	}
+
+	@Override
+	public String representation() {
+		return isTop() ? "Non-constant" : isBottom() ? "Bot" : value.toString();
+	}
+
+	@Override
+	protected ConstantPropagation evalNullConstant() {
+		return top();
+	}
+
+	@Override
+	protected ConstantPropagation evalNonNullConstant(Constant constant) {
+		if (constant.getValue() instanceof Integer)
+			return new ConstantPropagation((Integer) constant.getValue());
+		return top();
+	}
+
+	@Override
+	protected ConstantPropagation evalTypeConversion(Type type, ConstantPropagation arg) {
+		return top();
+	}
+
+	@Override
+	protected ConstantPropagation evalUnaryExpression(UnaryOperator operator, ConstantPropagation arg) {
+		switch (operator) {
+		case NUMERIC_NEG:
+			return new ConstantPropagation(0 - value);
+		case STRING_LENGTH:
+			return top();
+		default:
+			return top();
+		}
+	}
+
+	@Override
+	protected ConstantPropagation evalBinaryExpression(BinaryOperator operator, ConstantPropagation left,
+			ConstantPropagation right) {
+		switch (operator) {
+		case NUMERIC_ADD:
+			return new ConstantPropagation(left.value + right.value);
+		case NUMERIC_DIV:
+			if (left.value % right.value != 0)
+				return top();
+			else
+				return new ConstantPropagation(left.value / right.value);
+		case NUMERIC_MOD:
+			return new ConstantPropagation(left.value % right.value);
+		case NUMERIC_MUL:
+			return new ConstantPropagation(left.value * right.value);
+		case NUMERIC_SUB:
+			return new ConstantPropagation(left.value - right.value);
+		default:
+			return top();
+		}
+	}
+
+	@Override
+	protected ConstantPropagation evalTernaryExpression(TernaryOperator operator, ConstantPropagation left,
+			ConstantPropagation middle, ConstantPropagation right) {
+		return top();
+	}
+
+	@Override
+	protected ConstantPropagation lubAux(ConstantPropagation other) throws SemanticException {
+		return equals(other) ? other : top();
```

Comment:
No need of calling `equals()`: the `lub()` implementation in `BaseLattice` takes care of this case (if this is not explicit in the javadoc, state it)